src/ui/public/vislib/lib/axis.js

```diff
@@ -5,145 +5,113 @@ import ErrorHandlerProvider from 'ui/vislib/lib/_error_handler';
 import AxisTitleProvider from 'ui/vislib/lib/axis_title';
 import AxisLabelsProvider from 'ui/vislib/lib/axis_labels';
 import AxisScaleProvider from 'ui/vislib/lib/axis_scale';
+import AxisConfigProvider from 'ui/vislib/lib/axis_config';
 
 export default function AxisFactory(Private) {
   const ErrorHandler = Private(ErrorHandlerProvider);
   const AxisTitle = Private(AxisTitleProvider);
   const AxisLabels = Private(AxisLabelsProvider);
   const AxisScale = Private(AxisScaleProvider);
-  const defaults = {
-    show: true,
-    type: 'value',
-    elSelector: '.axis-wrapper-{pos} .axis-div',
-    position: 'left',
-    axisFormatter: null, // TODO: create default axis formatter
-    scale: 'linear',
-    expandLastBucket: true, //TODO: rename ... bucket has nothing to do with vis
-    inverted: false,
-    style: {
-      color: '#ddd',
-      lineWidth: '1px',
-      opacity: 1,
-      tickColor: '#ddd',
-      tickWidth: '1px',
-      tickLength: '6px'
-    }
-  };
-
-  const categoryDefaults = {
-    type: 'category',
-    position: 'bottom',
-    labels: {
-      rotate: 0,
-      rotateAnchor: 'end',
-      filter: true
-    }
-  };
-  /**
-   * Appends y axis to the visualization
-   *
-   * @class Axis
-   * @constructor
-   * @param args {{el: (HTMLElement), yMax: (Number), _attr: (Object|*)}}
-   */
+  const AxisConfig = Private(AxisConfigProvider);
+
   class Axis extends ErrorHandler {
     constructor(args) {
       super();
-      if (args.type === 'category') {
-        _.extend(this, defaults, categoryDefaults, args);
-      } else {
-        _.extend(this, defaults, args);
+      this.data = args.data;
+      this.config = new AxisConfig(args);
+      if (this.config.get('type') === 'category') {
+        this.values = this.data.xValues();
+        this.ordered = this.data.get('ordered');
       }
-
-      this._attr = args.vis._attr;
-      this.elSelector = this.elSelector.replace('{pos}', this.position);
-      this.scale = new AxisScale(this, {scale: this.scale});
-      this.axisTitle = new AxisTitle(this, this.axisTitle);
-      this.axisLabels = new AxisLabels(this, this.labels);
+      this.axisScale = new AxisScale(this.config, this.data);
+      this.axisTitle = new AxisTitle(this.config);
+      this.axisLabels = new AxisLabels(this.config, this.axisScale);
     }
 
     render() {
-      d3.select(this.vis.el).selectAll(this.elSelector).call(this.draw());
-    }
-
-    isHorizontal() {
-      return (this.position === 'top' || this.position === 'bottom');
+      const elSelector = this.config.get('elSelector');
+      const rootEl = this.config.get('rootEl');
```

Comment:
yeah,  i completely agree. however i think this should be changed at a later stage, when we start doing changes to the vis itself, for now i just made rootEl availible on config instead of going to vis.el